### ISSUE TITLE
Preferred word score candidate rejected on quick search

### ISSUE THREAD

**Issue by Beanow** [Sonarr/Sonarr#4977]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Searching for an episode manually/interactively,
![image](https://user-images.githubusercontent.com/497556/162257239-9d2597e3-49ee-4794-b34b-d73aa16cefd1.png)
A suitable candidate exists with +10 word score (based on the release group). And it shows no reason to reject it.
![image](https://user-images.githubusercontent.com/497556/162257361-0e259df5-263c-4435-a5b6-09ceb82afb0b.png)

However using an automatic/quick search, a release with a neutral score (not preferred) is grabbed.
![image](https://user-images.githubusercontent.com/497556/162258256-259602e1-0744-40a6-96a5-8aaaab9e844e.png)

A quick glance at the trace log **I think "current" and "new" are accidentally reversed**. Therefore rejecting the preferred one.
In the first place there shouldn't be a "current" as this episode wasn't grabbed/downloaded at all yet.
```
2022-04-07 19:04:02.5|Debug|UpgradableSpecification|Comparing preferred word score. Current: 10 New: 0
2022-04-07 19:04:02.5|Debug|UpgradableSpecification|Existing item meets cut-off. skipping.
...
2022-04-07 19:04:02.6|Debug|DownloadDecisionMaker|Release rejected for the following reasons: [Permanent] Existing file meets cutoff: HDTV-1080p - English
```

### Expected Behavior

When multiple candidates exist in an automatic/quick search, the highest preferred word score result should be selected.

### Steps To Reproduce

1. Setting up a release groups profile (see below) with preference scores.
2. Set up AnimeTosho / Nyaa indexers.
3. Add the example series.
4. Perform an interactive search to confirm *no rejection reasons* are given for a positive score release.
5. Don't start a download from interactive search.
6. Perform a quick search for the same episode.
7. Confirm in activity & trace logs, it rejects the positive score release(s).

### Environment

```markdown
- OS: Ubuntu 20.04 (docker host)
- Sonarr: 3.0.7.1477-ls140 by linuxserver.io
- Docker Install: Yes
- Using Reverse Proxy: Yes
- Browser: N/A
```

### What branch are you running?

Main

### Trace Logs?

Info log
```
2022-04-07 19:02:22.8|Info|OwinHostController|Listening on the following URLs:
2022-04-07 19:02:22.8|Info|OwinHostController|  http://*:8989/
2022-04-07 19:02:23.2|Info|SonarrBootstrapper|Starting Web Server
2022-04-07 19:03:24.9|Info|ReleaseSearchService|Searching indexers for [Attack on Titan : S04E27 (86)]. 2 active indexers
2022-04-07 19:04:01.6|Info|DownloadDecisionMaker|Processing 87 releases
2022-04-07 19:04:03.9|Info|DownloadService|Report sent to Deluge. [naiyas-weeklies] Attack on Titan - Shingeki no Kyojin (The Final Season Part 2) - S04E27[86] - (WEBDL 1080p HEVC - 10 AAC Multiple Subtitle).mkv
```

[sonarr.trace.txt](https://github.com/Sonarr/Sonarr/files/8445369/sonarr.trace.txt)

### Anything else?

Release groups profile
![image](https://user-images.githubusercontent.com/497556/162264714-3604eead-7b6b-40d4-b718-2ff102d0ac75.png)

**Comment by bakerboy448** [comment]
Interactive Search and Logs confirm the release was not rejected

```
2022-04-07 19:04:02.8|Debug|ReleaseRestrictionsSpecification|Checking if release meets restrictions: [SubsPlease] Shingeki no Kyojin (The Final Season) - 86 (1080p) [F045593A].mkv
2022-04-07 19:04:02.8|Debug|ReleaseRestrictionsSpecification|[[SubsPlease] Shingeki no Kyojin (The Final Season) - 86 (1080p) [F045593A].mkv] No restrictions apply, allowing
2022-04-07 19:04:02.8|Debug|RetentionSpecification|Not checking retention requirement for non-usenet report
2022-04-07 19:04:02.8|Debug|SceneMappingSpecification|No explicit scene origin in scene mapping.
2022-04-07 19:04:02.8|Debug|SeriesSpecification|Checking if series matches searched series
2022-04-07 19:04:02.8|Debug|DelaySpecification|Ignoring delay for user invoked search
2022-04-07 19:04:02.8|Debug|HistorySpecification|Skipping history check during search
2022-04-07 19:04:02.8|Debug|MonitoredEpisodeSpecification|Skipping monitored check during search
2022-04-07 19:04:02.8|Trace|ConfigService|Using default config value for 'autounmonitorpreviouslydownloadedepisodes' defaultValue:'False'
2022-04-07 19:04:02.8|Debug|DownloadDecisionMaker|Release accepted
```
vs. the grabbed release
```
2022-04-07 19:04:02.3|Debug|QualityParser|Trying to parse quality for [naiyas-weeklies] Attack on Titan - Shingeki no Kyojin (The Final Season Part 2) - S04E27[86] - (WEBDL 1080p HEVC - 10 AAC Multiple Subtitle).mkv
2022-04-07 19:04:02.3|Debug|Parser|Quality parsed: WEBDL-1080p v1
2022-04-07 19:04:02.3|Debug|Parser|Release Group parsed: naiyas-weeklies
2022-04-07 19:04:02.3|Debug|ParsingService|Found matching series by TVDB ID 267440, an alias may be needed for: Attack on Titan - Shingeki no Kyojin (The Final Season Part 2)
2022-04-07 19:04:02.3|Debug|ParsingService|Using absolute episode number 86 for: Attack on Titan - TVDB: 4x27
2022-04-07 19:04:02.3|Trace|PreferredWordService|Calculating preferred word score for '[naiyas-weeklies] Attack on Titan - Shingeki no Kyojin (The Final Season Part 2) - S04E27[86] - (WEBDL 1080p HEVC - 10 AAC Multiple Subtitle).mkv'
2022-04-07 19:04:02.3|Trace|PreferredWordService|Calculated preferred word score for '[naiyas-weeklies] Attack on Titan - Shingeki no Kyojin (The Final Season Part 2) - S04E27[86] - (WEBDL 1080p HEVC - 10 AAC Multiple Subtitle).mkv': 0 (0 match(es))
2022-04-07 19:04:02.3|Debug|AcceptableSizeSpecification|Beginning size check for: [naiyas-weeklies] Attack on Titan - Shingeki no Kyojin (The Final Season Part 2) - S04E27[86] - (WEBDL 1080p HEVC - 10 AAC Multiple Subtitle).mkv
2022-04-07 19:04:02.3|Debug|AcceptableSizeSpecification|Item: [naiyas-weeklies] Attack on Titan - Shingeki no Kyojin (The Final Season Part 2) - S04E27[86] - (WEBDL 1080p HEVC - 10 AAC Multiple Subtitle).mkv, meets size constraints
2022-04-07 19:04:02.3|Debug|AlreadyImportedSpecification|Performing already imported check on report
2022-04-07 19:04:02.3|Debug|AlreadyImportedSpecification|Skipping already imported check for episode without file
2022-04-07 19:04:02.3|Debug|LanguageSpecification|Checking if report meets language requirements. English
2022-04-07 19:04:02.3|Trace|ConfigService|Using default config value for 'maximumsize' defaultValue:'0'
2022-04-07 19:04:02.3|Debug|MaximumSizeSpecification|Maximum size is not set.
2022-04-07 19:04:02.3|Debug|MinimumAgeSpecification|Not checking minimum age requirement for non-usenet report
2022-04-07 19:04:02.3|Debug|QualityAllowedByProfileSpecification|Checking if report meets quality requirements. WEBDL-1080p v1
2022-04-07 19:04:02.3|Debug|ReleaseRestrictionsSpecification|Checking if release meets restrictions: [naiyas-weeklies] Attack on Titan - Shingeki no Kyojin (The Final Season Part 2) - S04E27[86] - (WEBDL 1080p HEVC - 10 AAC Multiple Subtitle).mkv
2022-04-07 19:04:02.3|Debug|ReleaseRestrictionsSpecification|[[naiyas-weeklies] Attack on Titan - Shingeki no Kyojin (The Final Season Part 2) - S04E27[86] - (WEBDL 1080p HEVC - 10 AAC Multiple Subtitle).mkv] No restrictions apply, allowing
2022-04-07 19:04:02.3|Debug|RetentionSpecification|Not checking retention requirement for non-usenet report
2022-04-07 19:04:02.3|Debug|SceneMappingSpecification|No applicable scene mapping, skipping.
2022-04-07 19:04:02.3|Debug|SeriesSpecification|Checking if series matches searched series
2022-04-07 19:04:02.3|Debug|DelaySpecification|Ignoring delay for user invoked search
2022-04-07 19:04:02.3|Debug|HistorySpecification|Skipping history check during search
2022-04-07 19:04:02.3|Debug|MonitoredEpisodeSpecification|Skipping monitored check during search
2022-04-07 19:04:02.3|Trace|ConfigService|Using default config value for 'autounmonitorpreviouslydownloadedepisodes' defaultValue:'False'
2022-04-07 19:04:02.3|Debug|DownloadDecisionMaker|Release accepted
```

Your screenshot shows 2 HDTV releases which are lower qualities than WEBDL and thus would be rejected.
https://wiki.servarr.com/sonarr/faq#how-are-possible-downloads-compared

Also why does Kodi need to know and rescan it's library when Sonarr sends a download to the client? That doesn't really make sense.


However, there's no bug as far as the logging goes - your file does exist on disk and is for a completely different episode.

Did you capture these logs in the middle of an RSS refresh or something? Or does the tracker just return every single release for the season?

```
2022-04-07 19:04:02.6|Trace|EpisodeFilePreferredWordCalculator|Calculated preferred word score for 'Season 4/[SubsPlease] Shingeki no Kyojin (The Final Season) - 87 (1080p) [30012FE7].mkv': 10 (1 match(es))
2022-04-07 19:04:02.6|Debug|UpgradableSpecification|Comparing preferred word score. Current: 10 New: 0
```

**Comment by markus101** [comment]
> A quick glance at the trace log I think "current" and "new" are accidentally reversed. Therefore rejecting the preferred one.

This does not appear to be the case:

```
2022-04-07 19:04:02.5|Trace|PreferredWordService|Calculating preferred word score for 'Attack.on.Titan.Shingeki.no.Kyojin.The.Final.Season.S04E27.E86.1080p.CR.WEBRip.10bits.x265-Rapta'
2022-04-07 19:04:02.5|Trace|PreferredWordService|Calculated preferred word score for 'Attack.on.Titan.Shingeki.no.Kyojin.The.Final.Season.S04E27.E86.1080p.CR.WEBRip.10bits.x265-Rapta': 0 (0 match(es))
2022-04-07 19:04:02.5|Debug|AcceptableSizeSpecification|Beginning size check for: Attack.on.Titan.Shingeki.no.Kyojin.The.Final.Season.S04E27.E86.1080p.CR.WEBRip.10bits.x265-Rapta
2022-04-07 19:04:02.5|Debug|AcceptableSizeSpecification|Item: Attack.on.Titan.Shingeki.no.Kyojin.The.Final.Season.S04E27.E86.1080p.CR.WEBRip.10bits.x265-Rapta, meets size constraints
2022-04-07 19:04:02.5|Debug|AlreadyImportedSpecification|Performing already imported check on report
2022-04-07 19:04:02.5|Debug|AlreadyImportedSpecification|Skipping already imported check for episode without file
2022-04-07 19:04:02.5|Debug|CutoffSpecification|Comparing file quality and language with report. Existing file is HDTV-1080p v1 - English
2022-04-07 19:04:02.5|Trace|EpisodeFilePreferredWordCalculator|Calculating preferred word score for '[SubsPlease] Shingeki no Kyojin (The Final Season) - 87 (1080p) [30012FE7]'
2022-04-07 19:04:02.5|Trace|EpisodeFilePreferredWordCalculator|Calculated preferred word score for '[SubsPlease] Shingeki no Kyojin (The Final Season) - 87 (1080p) [30012FE7]': 10 (1 match(es))
2022-04-07 19:04:02.5|Trace|EpisodeFilePreferredWordCalculator|Calculating preferred word score for '[SubsPlease] Shingeki no Kyojin (The Final Season) - 87 (1080p) [30012FE7]'
2022-04-07 19:04:02.5|Trace|EpisodeFilePreferredWordCalculator|Calculated preferred word score for '[SubsPlease] Shingeki no Kyojin (The Final Season) - 87 (1080p) [30012FE7]': 10 (1 match(es))
2022-04-07 19:04:02.5|Trace|EpisodeFilePreferredWordCalculator|Calculating preferred word score for 'Season 4/[SubsPlease] Shingeki no Kyojin (The Final Season) - 87 (1080p) [30012FE7].mkv'
2022-04-07 19:04:02.5|Trace|EpisodeFilePreferredWordCalculator|Calculated preferred word score for 'Season 4/[SubsPlease] Shingeki no Kyojin (The Final Season) - 87 (1080p) [30012FE7].mkv': 10 (1 match(es))
2022-04-07 19:04:02.5|Debug|UpgradableSpecification|Comparing preferred word score. Current: 10 New: 0
```

Looks like there is something funky with the mapping of your episode files, absolute episode number 86 vs 87

> In the first place there shouldn't be a "current" as this episode wasn't grabbed/downloaded at all yet.

There are several places where this same check is used, one is for the existing file, another for releases from the history (only for RSS Sync) and then items tracked in the queue, which would include things that are pending being grabbed, though in this case it's comparing to a file.

> Also why does Kodi need to know and rescan it's library when Sonarr sends a download to the client? That doesn't really make sense.

It doesn't, you can send toast messages to Kodi, which is what this would be.

**Comment by Beanow** [comment]
> Did you capture these logs in the middle of an RSS refresh or something? Or does the tracker just return every single release for the season?

The tracker releases a lot of episodes that weren't requested. Plus I have Nyaa and AnimeTosho enabled with overlapping results. So `Processing release 24/87` yep 87. Identical to what I see in interactive search.

I've only set kodi to notifications, not library updates.

> Looks like there is something funky with the mapping of your episode files, absolute episode number 86 vs 87

I did have 87 already downloaded when searching for 86. But 86 wasn't downloading / present, it was in wanted.
![image](https://user-images.githubusercontent.com/497556/162271000-91644d0c-6676-41f6-9988-ba9332dccdc4.png)
